File: 90-subsets-ii/subsets-ii.cpp
class Solution {
public:
    void Recursivesubsets(vector<int>& nums, vector<vector<int>> &allsubsets, vector<int> &ans, int i){
        if(i == nums.size()){
            allsubsets.push_back(ans);
            return;
        }
        ans.push_back(nums[i]);
        Recursivesubsets(nums, allsubsets, ans, i + 1);

        ans.pop_back();
        int idx = i + 1;
        while(idx < nums.size() && nums[idx] == nums[idx - 1]){
            idx++;
        }
        Recursivesubsets(nums, allsubsets, ans, idx);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> allsubsets;
        vector<int> ans;
        Recursivesubsets(nums, allsubsets, ans, 0);
        return allsubsets;
    }
};

File: 90-subsets-ii/subsets-ii.py
class Solution:
    def recursive_subsets(self, nums: list[int], all_subsets: list[list[int]], ans: list[int], i: int) -> None:
        if i == len(nums):
            all_subsets.append(ans.copy())
            return

        ans.append(nums[i])
        self.recursive_subsets(nums, all_subsets, ans, i + 1)

        ans.pop()
        idx = i + 1
        while idx < len(nums) and nums[idx] == nums[idx - 1]:
            idx += 1
        self.recursive_subsets(nums, all_subsets, ans, idx)

    def subsetsWithDup(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        all_subsets = []
        ans = []
        self.recursive_subsets(nums, all_subsets, ans, 0)
        return all_subsets